### ISSUE TITLE
Vessels pull away from planet during physical timewarp

### ISSUE THREAD

**Issue by maccollo** [mockingbirdnest/Principia#1307]
It appears that vessels with running engines want to orient away from the planet.
This behavior stops when one forces principia to stop via the UI.
The effect appears to be proportional to the amount of throttle (maybe acceleration?), and does not appear to happen when physical time acceleration is at 1x. It also does not happen when physics warping with no engines running.

https://www.youtube.com/watch?v=w72VDoG87Yg

**Comment by Damien212** [comment]
Wall of text copied from the forum (looks like same bug):

While playing RSS/RO with Principia (latest releases as of today), during a mission similar to Explorer I (rocket booster stage, coast to apo, chained SRB to orbital velocity) the SRB stage was suffering from what looked like misaligned thrust, pulling the craft from orbital prograde, towards radial-out with reference to Earth. While trying to diagnose the problem I tried the following:

Delete/Fresh install of KSP, with RSS, RO & dependancies as per CKAN latest. Also FASA for test craft. Principia as per latest on github. - no change
Spin stablization of SRB stage - no effect. craft nose still pulled towards radial-out causing spin.
Disabling/reenabling of RCS/SAS - no effect (not installed on craft anyway)
SRB stage activated while pointing retrograde - no effect, nose still pulled towards radial-out causing spin.
SRB stage activated while pointing radial-out - no alteration of vessel orientation, remains pointing radial-out
Craft (FASA Juno/Explorer I craft file sergeant SRB stages as per github) replaced with alternate model to rule out FASA part issue - no effect, procedural parts SRB with a MK1 crew capsule on top exhibits same issues
Noticed Moon was directly above launch site, timewarped Moon to other side of Earth before launch to rule out weird gravity effects from Moon - no effect.
Force stopped Principia before stage ignition - issue eliminated, works as expected
Force started Principia again before stage ignition - issue returns.
The only constant I can find regarding this issue is that it only occurs when thrust is applied out of atmo. At 139,999m everything works as normal, at 140,000m+ the problem appears. If you start a burn in atmo, and the burn continues across the boundary, the issue takes effect only from the moment the craft passes the boundary.

The craft I was testing has 3x SRB stages to push Explorer I into orbit. During the first burn towards orbital prograde, the craft noses upwards, towards radial-out and then towards orbital retrograde, putting it into an anticlockwise spin if viewed from south to north. If this was asymmetric thrust or a simple spin, you'd expect the spin to continue in the same direction if you ignited the second stage when pointing retrograde. This doesn't happen. Instead you can start the spin anticlockwise with stage 1 ignition when facing prograde, ignite stage 2 during the spin when at retrograde, which changes the direction of the spin to clockwise towards radial-out and prograde, then ignite stage 3 at prograde again to reverse direction. In all cases the craft spins towards radial-out during thrust.

 

I recorded several attempts at performing the maneuver as per the above list. The video is here: https://www.twitch.tv/videos/148710758

My KSP.log is here: https://www.dropbox.com/s/0tvzdx4spaw0lm7/KSP.log?dl=0

**Comment by eggrobin** [comment]
This seems to occur at large timesteps or at large distances from the planet, and it is a rotation of the parts (something that should not happen, we should move things in a rigid way) biased away from the planet; I am wondering whether it might be cancellations due to `PileUp` and related classes not using `DoublePrecision` double-double...

**Comment by eggrobin** [comment]
Some experimental results: this does not appear to require either physics warp or RSS, any unsteered thrust will exhibit this; it takes more time to drift to the radial direction at lower thrusts.

**Comment by eggrobin** [comment]
This occurs even on single-part vessels (tested with a Mk 2 Pod in RO using its built-in RCS.

**Comment by eggrobin** [comment]
Ok, the explanation is that when an EVA Kerbal is created, it ignores collisions with the ship that spawned it for a little while. This appears to be done by nudging it back in place in the same way we nudge things, after yielding WaitForFixedUpdate; since we expect nothing to run there, random things happen.
Since KSP is taking care of moving that Kerbal around, we don't need to move it, so this is solved by adding an unmanageability reason:
```c#
    if (vessel.isEVA && !vessel.evaController.Ready) {
      reasons.Add("vessel is an unready Kerbal");
    }
```
The Kerbal becomes known to Principia and is nudged exclusively by us once it is ready.

I will make a CL tomorrow to fix this.

**Comment by eggrobin** [comment]
A note two years later, since I came across this fix when considering handling rotations in the pile up: the miraculous fix in 1e4547a59463384bc9fca852fce9141eca54e7b0,
https://github.com/mockingbirdnest/Principia/blob/1e4547a59463384bc9fca852fce9141eca54e7b0/ksp_plugin_adapter/ksp_plugin_adapter.cs#L1104-L1105
after setting `part.rb.position` to the same, is probably explained by the documented behaviour of `RigidBody`.

The [documentation](https://docs.unity3d.com/ScriptReference/Rigidbody-position.html) for `RigidBody.position` says

> Rigidbody.position allows you to get and set the position of a Rigidbody using the physics engine. If you change the position of a Rigibody using Rigidbody.position, the transform will be updated after the next physics simulation step.

Setting `part.rb.position` after the physics step (after yielding `WaitForFixedUpdate`) would thus only set the position at the next step (and the position being off-by-one in motion around the Earth can probably explain a torque righting the thrust away from it).
